src/Raven.Client/Documents/Session/AsyncSessionDocumentTimeSeries.cs

```diff
@@ -43,27 +88,344 @@ public async Task<IEnumerable<TimeSeriesValue>> GetAsync(string timeseries, Date
+            DateTime from, 
+            DateTime to, 
+            TimeSeriesRange fromRange, 
+            TimeSeriesRange toRange, 
+            List<TimeSeriesValue> values, 
+            int skip, 
+            int trim)
+        {
+            if (fromRange != null && fromRange.To >= from)
+            {
+                // need to skip a part of the first range 
+
+                if (toRange != null && toRange.From <= to)
                 {
-                    //TODO: if value exists, merge new values
-                    Session.TimeSeriesByDocId[DocId] = details;
+                    // also need to trim a part of the last range 
+                    return values.Skip(skip).Take(values.Count - skip - trim);
                 }
+
+                return values.Skip(skip);
             }
-            //TODO: Make sure that we get just the relevant range
-            return details.Values[timeseries].Values;
 
+            if (toRange != null && toRange.From <= to)
+            {
+                // trim a part of the last range 
+                return values.Take(values.Count - trim);
+            }
+
+            return values;
+        }
+
+        private static bool GetFromCacheOrFigureOutMissingParts(
+            DateTime from, 
+            DateTime to,
+            List<TimeSeriesRange> ranges, 
+            out int fromRangeIndex, 
+            out int toRangeIndex, 
+            out List<(DateTime, DateTime)> rangesToGetFromServer,
+            out IEnumerable<TimeSeriesValue> result)
+        {
+            // try to find a range in cache that contains [from, to]
+            // if found, chop just the relevant part from it and return to the user.
+
+            // otherwise, try to find two ranges (fromRange, toRange),
+            // such that 'fromRange' is the last occurence for which range.From <= from
+            // and 'toRange' is the first occurence for which range.To >= to.
+            // At the same time, figure out the missing partial ranges that we need to get from the server.
+
+            fromRangeIndex = -1;
+            rangesToGetFromServer = default;
+
+            for (toRangeIndex = 0; toRangeIndex < ranges.Count; toRangeIndex++)
+            {
+                if (ranges[toRangeIndex].From <= from)
+                {
+                    if (ranges[toRangeIndex].To >= to)
+                    {
+                        // we have the entire range in cache
+
+                        result = ChopRelevantRange(ranges[toRangeIndex], from, to);
+                        return true;                     
+                    }
+
+                    fromRangeIndex = toRangeIndex;
+                    continue;
+                }
+
+                rangesToGetFromServer = rangesToGetFromServer ?? new List<(DateTime, DateTime)>();
+
+                var (f, t) = (toRangeIndex == 0 || ranges[toRangeIndex - 1].To < from ? from : ranges[toRangeIndex - 1].To,
```

Comment:
This is dense code, can you break it up?

Reply:
done